### PR TITLE
fix: add background energy virial correction for non-neutral Ewald/PME systems

### LINKED ISSUE #61: 🐛[BUG]: particle_mesh_ewald explicit virial missing E_background gradient for non-neutral systems
### Version

0.3.0

### On which installation method(s) does this occur?

_No response_

### Describe the issue

The explicit virial returned by `particle_mesh_ewald(..., compute_virial=True)` is missing the gradient contribution from the background energy term `E_bg = pi / (2 * alpha^2 * V) * Q^2`. This term only appears for non-neutral systems (`Q_total != 0`), so the discrepancy is not caught by existing tests which use near-neutral charges (`Q = +0.5`) with high tolerance.

### Minimum reproducible example

```python
# %%
# Versions
# ===========================
import math

import nvalchemiops
import torch
from nvalchemiops.torch.interactions.electrostatics import (
    estimate_pme_parameters,
    particle_mesh_ewald,
)
from nvalchemiops.torch.neighbors import neighbor_list

print("torch.__version__:", torch.__version__)
print("nvalchemiops.__version__:", nvalchemiops.__version__)
"""
torch.__version__: 2.8.0
nvalchemiops.__version__: 0.3.0
"""

# %%
# Setup
# ===========================
#
# The explicit virial returned by particle_mesh_ewald does not match the
# virial computed by differentiating the energy w.r.t. strain.
#
# We demonstrate this with three independent virial computations:
# 1. Finite-difference virial (ground truth, no autograd involved)
# 2. Autograd virial (differentiating energy w.r.t. a strain parameter)
# 3. Explicit virial (from particle_mesh_ewald's compute_virial=True)
#
# Methods 1 and 2 agree to machine precision. Method 3 does not.
# The discrepancy is diagonal and position-dependent, suggesting a missing contribution.

device = "cpu"
dtype = torch.float64

# Charged NaCl pair, so E_bg ≠ 0
positions = torch.tensor([[2.0, 3.0, 4.0], [7.0, 6.0, 8.0]], dtype=dtype, device=device)
charges = torch.tensor([1.0, -3.0], dtype=dtype, device=device)
cell = torch.tensor(
    [[[10.0, 0.5, 0.3], [0.5, 11.0, 0.2], [0.3, 0.2, 12.0]]], dtype=dtype, device=device
)

# Estimate PME parameters (fixed for all three methods)
n_total = positions.shape[0]
params = estimate_pme_parameters(
    positions,
    cell,
    batch_idx=torch.zeros(n_total, dtype=torch.int32, device=device),
    accuracy=1e-6,
)
alpha = params.alpha
cutoff = params.real_space_cutoff.max().item()
mesh_dimensions = tuple(params.mesh_dimensions)

neighbor_matrix, num_neighbors, neighbor_shift_matrix = neighbor_list(
    positions=positions,
    cell=cell,
    pbc=torch.tensor([[True, True, True]], device=device),
    cutoff=cutoff,
    batch_idx=torch.zeros(n_total, dtype=torch.int32, device=device),
    fill_value=n_total,
)

pme_kwargs = dict(
    alpha=alpha,
    mesh_dimensions=mesh_dimensions,
    spline_order=4,
    batch_idx=None,
    neighbor_matrix=neighbor_matrix.to(torch.int32),
    neighbor_matrix_shifts=neighbor_shift_matrix.to(torch.int32),
    mask_value=n_total,
    accuracy=1e-6,
)


# %%
# Method 1: Finite-difference virial (ground truth)
# ===========================

delta = 1e-6
eye = torch.eye(3, dtype=dtype, device=device)
fd_virial = torch.zeros(3, 3, dtype=dtype, device=device)

for i in range(3):
    for j in range(3):
        for sign in [+delta, -delta]:
            strain = eye.clone()
            strain[i, j] += sign
            pos_s = positions @ strain.T
            cell_s = (cell.squeeze(0) @ strain.T).unsqueeze(0)

            # Rebuild neighbor list for perturbed geometry
            nm_s, _, ns_s = neighbor_list(
                positions=pos_s,
                cell=cell_s,
                pbc=torch.tensor([[True, True, True]]),
                cutoff=cutoff,
                batch_idx=torch.zeros(n_total, dtype=torch.int32),
                fill_value=n_total,
            )

            e_s = particle_mesh_ewald(
                positions=pos_s,
                charges=charges,
                cell=cell_s,
                compute_forces=False,
                compute_charge_gradients=False,
                compute_virial=False,
                alpha=alpha,
                mesh_dimensions=mesh_dimensions,
                spline_order=4,
                batch_idx=None,
                neighbor_matrix=nm_s.to(torch.int32),
                neighbor_matrix_shifts=ns_s.to(torch.int32),
                mask_value=n_total,
                accuracy=1e-6,
            )
            if sign > 0:
                e_p = e_s.sum()
            else:
                e_m = e_s.sum()

        fd_virial[i, j] = -(e_p - e_m) / (2 * delta)


# %%
# Method 2: Autograd virial (via strain)
# ===========================

strain = torch.zeros(3, 3, dtype=dtype, device=device, requires_grad=True)
deformation = eye + strain
strained_positions = positions @ deformation.T
strained_cell = (cell.squeeze(0) @ deformation.T).unsqueeze(0)

per_atom_energies = particle_mesh_ewald(
    positions=strained_positions,
    charges=charges,
    cell=strained_cell,
    compute_forces=False,
    compute_charge_gradients=False,
    compute_virial=False,
    **pme_kwargs,
)
energy = per_atom_energies.sum()
autograd_virial = -torch.autograd.grad(energy, strain)[0]


# %%
# Method 3: Explicit virial from particle_mesh_ewald
# ===========================

_, _, explicit_virial = particle_mesh_ewald(
    positions=positions,
    charges=charges,
    cell=cell,
    compute_forces=True,
    compute_charge_gradients=False,
    compute_virial=True,
    **pme_kwargs,
)
explicit_virial = explicit_virial.squeeze(0)


# %%
# Compare
# ===========================

print("Method 1 — FD virial (ground truth):")
print(fd_virial)
print()
print("Method 2 — Autograd virial (energy + strain):")
print(autograd_virial)
print()
print("Method 3 — Explicit virial (particle_mesh_ewald):")
print(explicit_virial)
print()

fd_vs_autograd = (fd_virial - autograd_virial).abs().max().item()
fd_vs_explicit = (fd_virial - explicit_virial).abs().max().item()
autograd_vs_explicit = (autograd_virial - explicit_virial).abs().max().item()

print(f"FD vs autograd:  {fd_vs_autograd:.2e}  (should be ~0)")
print(f"FD vs explicit:  {fd_vs_explicit:.2e}  (should be ~0, but ISN'T)")
print(f"Autograd vs explicit: {autograd_vs_explicit:.2e}")
print()

diff = fd_virial - explicit_virial
print("Difference (FD - explicit):")
print(diff)
print()
print(f"Diagonal:     [{diff[0, 0].item():.6f}, {diff[1, 1].item():.6f}, {diff[2, 2].item():.6f}]")
print(f"Off-diagonal max: {(diff - torch.diag(diff.diag())).abs().max().item():.2e}")
"""
Method 1 — FD virial (ground truth):
tensor([[-0.3239,  0.0245,  0.0296],
        [ 0.0245, -0.2789, -0.0531],
        [ 0.0296, -0.0531, -0.5272]], dtype=torch.float64)

Method 2 — Autograd virial (energy + strain):
tensor([[-0.3239,  0.0245,  0.0296],
        [ 0.0245, -0.2789, -0.0531],
        [ 0.0296, -0.0531, -0.5272]], dtype=torch.float64)

Method 3 — Explicit virial (particle_mesh_ewald):
tensor([[-0.1790,  0.0245,  0.0296],
        [ 0.0245, -0.1340, -0.0531],
        [ 0.0296, -0.0531, -0.3823]], dtype=torch.float64)

FD vs autograd:  5.64e-10  (should be ~0)
FD vs explicit:  1.45e-01  (should be ~0, but ISN'T)
Autograd vs explicit: 1.45e-01

Difference (FD - explicit):
tensor([[-1.4487e-01, -1.3050e-07, -7.1850e-09],
        [-1.3094e-07, -1.4487e-01, -5.2153e-07],
        [-7.9622e-09, -5.2131e-07, -1.4487e-01]], dtype=torch.float64)

Diagonal:     [-0.144868, -0.144867, -0.144867]
Off-diagonal max: 5.22e-07
"""
# %%
# Correct explicit virial by adding missing term
# ===========================
#
# Background virial from E_bg = π/(2α²V) · Q². Since E_total subtracts
# E_bg, the virial contribution is W_bg = dE_bg/dε = -E_bg · I.
# This term is missing from the explicit virial entirely.


volume = torch.det(cell.squeeze(0)).abs()
Q_total = charges.sum()
alpha_val = alpha.item() if alpha.numel() == 1 else alpha
E_bg = (math.pi / (2.0 * alpha_val**2)) * Q_total**2 / volume

corrected_virial = explicit_virial - E_bg * eye

diff_corrected = autograd_virial - corrected_virial
print()
print("=== With background correction (W_bg = -E_bg · I) ===")
print("Corrected virial:")
print(corrected_virial)
print()
print("Difference (autograd - corrected):")
print(diff_corrected)
print()
print(f"Max diff after correction: {diff_corrected.abs().max().item():.2e}")
"""
=== With background correction (W_bg = -E_bg · I) ===
Corrected virial:
tensor([[-0.3239,  0.0245,  0.0296],
        [ 0.0245, -0.2789, -0.0531],
        [ 0.0296, -0.0531, -0.5272]], dtype=torch.float64)

Difference (autograd - corrected):
tensor([[-8.3954e-07, -1.3055e-07, -7.3983e-09],
        [-1.3055e-07,  2.8332e-08, -5.2138e-07],
        [-7.3983e-09, -5.2138e-07,  5.9188e-07]], dtype=torch.float64)

Max diff after correction: 8.40e-07
"""
```

### Relevant log output

```shell
torch.__version__: 2.10.0+cu128
nvalchemiops.__version__: 0.3.0
Method 1 — FD virial (ground truth):
tensor([[-0.3239,  0.0245,  0.0296],
        [ 0.0245, -0.2789, -0.0531],
        [ 0.0296, -0.0531, -0.5272]], dtype=torch.float64)

Method 2 — Autograd virial (energy + strain):
tensor([[-0.3239,  0.0245,  0.0296],
        [ 0.0245, -0.2789, -0.0531],
        [ 0.0296, -0.0531, -0.5272]], dtype=torch.float64)

Method 3 — Explicit virial (particle_mesh_ewald):
tensor([[-0.1790,  0.0245,  0.0296],
        [ 0.0245, -0.1340, -0.0531],
        [ 0.0296, -0.0531, -0.3823]], dtype=torch.float64)

FD vs autograd:  5.64e-10  (should be ~0)
FD vs explicit:  1.45e-01  (should be ~0, but ISN'T)
Autograd vs explicit: 1.45e-01

Difference (FD - explicit):
tensor([[-1.4487e-01, -1.3050e-07, -7.1850e-09],
        [-1.3094e-07, -1.4487e-01, -5.2153e-07],
        [-7.9622e-09, -5.2131e-07, -1.4487e-01]], dtype=torch.float64)

Diagonal:     [-0.144868, -0.144867, -0.144867]
Off-diagonal max: 5.22e-07

=== With background correction (W_bg = -E_bg · I) ===
Corrected virial:
tensor([[-0.3239,  0.0245,  0.0296],
        [ 0.0245, -0.2789, -0.0531],
        [ 0.0296, -0.0531, -0.5272]], dtype=torch.float64)

Difference (autograd - corrected):
tensor([[-8.3954e-07, -1.3055e-07, -7.3983e-09],
        [-1.3055e-07,  2.8332e-08, -5.2138e-07],
        [-7.3983e-09, -5.2138e-07,  5.9188e-07]], dtype=torch.float64)

Max diff after correction: 8.40e-07
```

### Environment details

```shell
- **nvalchemiops**: 0.3.0
- **PyTorch**: 2.10.0+cu128
- **CUDA**: 12.8
- **Python**: 3.12.3
- **OS**: Linux 6.14.0
- **Reproduced on**: CPU
```

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit-Ops Pull Request

## Description

Fix missing background energy virial contribution in PME and Ewald reciprocal space calculations for non-neutral systems. The background energy term E_bg = pi * Q^2 / (2 * alpha^2 * V) was correctly subtracted from the energy but its derivative with respect to strain was not included in the explicit virial tensor, causing a discrepancy between explicit virial and finite-difference / autograd virial for charged systems.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

Fixes #61

## Changes Made

- Add background virial correction W_bg = -E_bg * I to PME reciprocal space virial in both PyTorch and JAX backends, handling single-system and batch cases
- Add the same background virial correction to all six Ewald reciprocal space variants in PyTorch (energy-only, energy+forces, energy+forces+charge_grad, and their batch counterparts) and both JAX variants (single and batch)
- Tighten non-neutral virial test tolerances from atol/rtol=2e-2 to 1e-4 to reflect the corrected implementation
- Increase charge imbalance in `make_non_neutral_system` from [1.0, -0.5] to [1.0, -3.5] for stronger test coverage and remove redundant high-charge test cases

## Testing

- [ ] Unit tests pass locally (`make pytest`)
- [ ] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

Existing non-neutral virial finite-difference tests (`test_ewald_total_virial_fd_non_neutral`, `test_pme_reciprocal_virial_fd_non_neutral`) now pass with 100x tighter tolerances (1e-4 vs 2e-2). No new tests needed since the existing tests already cover the affected code paths with the updated charge imbalance.

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

The mathematical derivation: E_bg = pi * Q_total^2 / (2 * alpha^2 * V) is subtracted from the reciprocal energy for non-neutral systems. Since V scales as V * det(I + epsilon) under strain, dV/d(epsilon) = V * I, giving dE_bg/d(epsilon) = -E_bg * I. The virial contribution is therefore W_bg = -E_bg * I, which must be subtracted from the reciprocal virial tensor.

Files changed:
- `nvalchemiops/torch/interactions/electrostatics/pme.py` - PyTorch PME virial fix
- `nvalchemiops/torch/interactions/electrostatics/ewald.py` - PyTorch Ewald virial fix (6 functions)
- `nvalchemiops/jax/interactions/electrostatics/pme.py` - JAX PME virial fix
- `nvalchemiops/jax/interactions/electrostatics/ewald.py` - JAX Ewald virial fix (2 paths)
- `test/interactions/electrostatics/bindings/torch/test_utils.py` - stronger charge imbalance
- `test/interactions/electrostatics/bindings/torch/test_ewald.py` - tightened tolerances
- `test/interactions/electrostatics/bindings/torch/test_pme.py` - tightened tolerances

### CODE REVIEW

**Review by laserkelvin — APPROVED** [review]
LGTM - do we need to update the sphinx docs with the addition at all? Just to document the formula for the contribution

**Review by laserkelvin — APPROVED** [review]
